### PR TITLE
Implement external outputs by allowing to add listeners for every output

### PR DESCRIPTION
**Pull request by Zanthelo**

### CODE REVIEW

File: source/yats/include/yats/task_configurator.h [hunk 1]
```diff
@@ -57,14 +58,22 @@ class task_configurator : public abstract_task_configurator
         return find<typename helper::output_tuple, std::tuple_element_t<index, type>>(m_outputs, Id);
     }
 
+    template <uint64_t Id, typename LambdaTask>
+    void add_listener(LambdaTask task)
```
Comment (Highlife1911) [inline source/yats/include/yats/task_configurator.h]: This is not a LambdaTask. Better name would be Callable: http://en.cppreference.com/w/cpp/concept/Callable

File: source/yats/include/yats/task_container.h [hunk 2]
```diff
@@ -107,31 +107,29 @@ class task_container : public abstract_task_container
         throw std::runtime_error("A not copyable type cannot be used in multiple connections.");
     }
 
-    template <size_t index = 0, typename T = typename helper::output_type, typename Output = std::enable_if_t<std::is_same<T, void>::value, T>>
-    std::enable_if_t<(index < helper::output_count)> write(Output output)
+    template <size_t Index = 0, typename T = typename helper::output_type>
+    std::enable_if_t<(Index < helper::output_count)> write(T output)
```
Comment (Highlife1911) [inline source/yats/include/yats/task_container.h]: Rename T to Output or something similar.

File: source/yats/include/yats/task_container.h [hunk 3]
```diff
@@ -107,31 +107,29 @@ class task_container : public abstract_task_container
...
     {
-        auto& value = std::get<index>(output);
-        const auto& following_nodes = std::get<index>(m_output);
-        for (size_t i = 0; i < following_nodes.size(); ++i)
+        // Contains the callbacks to write into inputs of the following tasks.
+        const auto& callbacks = std::get<Index>(m_output);
+        auto& slot = std::get<Index>(output);
+
+        // cend() - 1 and back() will fail for empty callbacks
```
Comment (Highlife1911) [inline source/yats/include/yats/task_container.h]: I think this is obvious. I would not make such a comment.

Comment (mflueggen) [reply source/yats/include/yats/task_container.h]: But it does not hurt.

**Pull request merged**